server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/fileformats/data/StarTreeDataWriter.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.compositeindex.datacube.startree.fileformats.data;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.store.IndexOutput;
+import org.opensearch.index.compositeindex.datacube.startree.node.InMemoryTreeNode;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+
+import static org.opensearch.index.compositeindex.CompositeIndexConstants.COMPOSITE_FIELD_MARKER;
+import static org.opensearch.index.compositeindex.datacube.startree.fileformats.StarTreeWriter.VERSION_CURRENT;
+import static org.opensearch.index.compositeindex.datacube.startree.node.FixedLengthStarTreeNode.SERIALIZABLE_DATA_SIZE_IN_BYTES;
+import static org.opensearch.index.compositeindex.datacube.startree.utils.StarTreeUtils.ALL;
+
+/**
+ * Utility class for serializing a star-tree data structure.
+ *
+ * @opensearch.experimental
+ */
+public class StarTreeDataWriter {
+
+    private static final Logger logger = LogManager.getLogger(StarTreeDataWriter.class);
+
+    /**
+     * Writes the star-tree data structure.
+     *
+     * @param indexOutput the IndexOutput to write the star-tree data
+     * @param rootNode    the root node of the star-tree
+     * @param numNodes    the total number of nodes in the star-tree
+     * @param name        the name of the star-tree field
+     * @return the total size in bytes of the serialized star-tree data
+     * @throws IOException if an I/O error occurs while writing the star-tree data
+     */
+    public static long writeStarTree(IndexOutput indexOutput, InMemoryTreeNode rootNode, int numNodes, String name) throws IOException {
```

Comment:
Same comment for static here and other methods in write flow.

These are not utils, these are integral business logic for writing star tree file formats